nova_vm/src/ecmascript/builtins/indexed_collections/typed_array_objects/typed_array_intrinsic_object.rs

```diff
@@ -595,13 +593,276 @@ impl TypedArrayPrototype {
     ) -> JsResult<Value> {
-        todo!()
+        let separator = arguments.get(0);
+        // 1. Let O be the this value.
+        let o = this_value;
+        // 2. Let taRecord be ? ValidateTypedArray(O, seq-cst).
+        let ta_record = validate_typed_array(agent, gc.nogc(), o, Ordering::SeqCst)?;
+        let o = TypedArray::try_from(o).unwrap();
+        // 3. Let len be TypedArrayLength(taRecord).
+        let (len, element_size) = match o {
+            TypedArray::Int8Array(_) => (
+                typed_array_length::<i8>(agent, &ta_record),
+                std::mem::size_of::<i8>(),
+            ),
+            TypedArray::Uint8Array(_) => (
+                typed_array_length::<u8>(agent, &ta_record),
+                std::mem::size_of::<u8>(),
+            ),
+            TypedArray::Uint8ClampedArray(_) => (
+                typed_array_length::<U8Clamped>(agent, &ta_record),
+                std::mem::size_of::<U8Clamped>(),
+            ),
+            TypedArray::Int16Array(_) => (
+                typed_array_length::<i16>(agent, &ta_record),
+                std::mem::size_of::<i16>(),
+            ),
+            TypedArray::Uint16Array(_) => (
+                typed_array_length::<u16>(agent, &ta_record),
+                std::mem::size_of::<u16>(),
+            ),
+            TypedArray::Int32Array(_) => (
+                typed_array_length::<i32>(agent, &ta_record),
+                std::mem::size_of::<i32>(),
+            ),
+            TypedArray::Uint32Array(_) => (
+                typed_array_length::<u32>(agent, &ta_record),
+                std::mem::size_of::<u32>(),
+            ),
+            TypedArray::BigInt64Array(_) => (
+                typed_array_length::<i64>(agent, &ta_record),
+                std::mem::size_of::<i64>(),
+            ),
+            TypedArray::BigUint64Array(_) => (
+                typed_array_length::<u64>(agent, &ta_record),
+                std::mem::size_of::<u64>(),
+            ),
+            TypedArray::Float32Array(_) => (
+                typed_array_length::<f32>(agent, &ta_record),
+                std::mem::size_of::<f32>(),
+            ),
+            TypedArray::Float64Array(_) => (
+                typed_array_length::<f64>(agent, &ta_record),
+                std::mem::size_of::<f64>(),
+            ),
+        };
+        // 4. If separator is undefined, let sep be ",".
+        let (sep_string, recheck_buffer) = if separator.is_undefined() {
+            (String::from_small_string(","), false)
+        } else if let Ok(sep) = String::try_from(separator) {
+            (sep, false)
+        } else {
+            // 5. Else, let sep be ? ToString(separator).
+            (
+                to_string(agent, gc.reborrow(), separator)?
+                    .unbind()
+                    .bind(gc.nogc()),
+                true,
+            )
+        };
+        if len == 0 {
+            return Ok(String::EMPTY_STRING.into_value());
+        }
```

Comment:
**Nitpick:** Couldn't this be moved to before step 4?

**Suggestion:** A fast-path could also be added for `len == 1`, then we wouldn't need to do all of the allocation for joining. But I guess that would be a very uncommon case, so might not be worth it.

Reply:
Nope, there's a throw-path in `? ToString(separator)` :(

Reply:
Damn, right...